core/modulesource.go

```diff
@@ -256,6 +271,203 @@ func (src *ModuleSource) AutomaticGitignore(ctx context.Context) (*bool, error)
 }
 
+// LoadContext loads a directory from the module context directory.
+//
+// If the module is local, it will load the directory from the local source
+// directly from the host.
+// In that case, the path is first resolved based on the caller's host location.
+// Then if the path is absolute, it will be relative to the context directory.
+// Otherwise, it will be relative to the module root directory.
+//
+// If the module is git, it will load the directory from the git repository
+// using its context directory.
+func (src *ModuleSource) LoadContext(ctx context.Context, dag *dagql.Server, path string, ignore []string) (inst dagql.Instance[*Directory], err error) {
+	// We exclude .git by default because it's not useful and tends to invalidate the cache.
+	excludes := []string{"**/.git"}
+	includes := []string{}
+
+	for _, p := range ignore {
+		if strings.HasPrefix(p, "!") {
+			includes = append(includes, p[1:])
+		} else {
+			excludes = append(excludes, p)
+		}
+	}
+
+	switch src.Kind {
+	case ModuleSourceKindLocal:
+		bk, err := src.Query.Buildkit(ctx)
+		if err != nil {
+			return inst, fmt.Errorf("failed to get buildkit api: %w", err)
+		}
+
+		// Retrieve the absolute path to the context directory (.git or dagger.json)
+		// and the module root directory (dagger.json)
+		ctxPath, modPath, err := src.ResolveContextPathFromModule(ctx)
+		if err != nil {
+			return inst, fmt.Errorf("failed to resolve context path: %w", err)
+		}
+
+		// If path is not absolute, it's relative to the module root directory.
+		// If path is absolute, it's relative to the context directory.
+		if !filepath.IsAbs(path) {
+			path = filepath.Join(modPath, path)
+		} else {
+			path = filepath.Join(ctxPath, path)
+		}
+
+		// We just check if the path is relative to the context directory,
+		// if not, that means it's a path that target an outside directory
+		// which is not allowed.
+		relativePathToCtx, err := filepath.Rel(ctxPath, path)
+		if err != nil {
+			return inst, fmt.Errorf("failed to get relative path to context: %w", err)
+		}
+
+		// If the relative path is outisde of the context directory, throw an error.
+		if strings.HasPrefix(relativePathToCtx, "..") {
+			return inst, fmt.Errorf("path %q is outside of context directory %q, path should be relative to the context directory", path, ctxPath)
+		}
+
+		_, desc, err := bk.LocalImport(ctx, src.Query.Platform().Spec(), path, excludes, includes)
+		if err != nil {
+			return inst, fmt.Errorf("failed to import local module src: %w", err)
+		}
+
+		inst, err = LoadBlob(ctx, dag, desc)
+		if err != nil {
+			return inst, fmt.Errorf("failed to load local module src: %w", err)
+		}
+
+		return inst, nil
+	case ModuleSourceKindGit:
+		slog.Debug("moduleSource.LoadContext: loading contextual directory from git", "path", path, "kind", src.Kind, "repo", src.AsGitSource.Value.HTMLURL())
+
+		// Use the Git context directory.
+		ctxDir := src.AsGitSource.Value.ContextDirectory.Self
+
```

Comment:
Shouldn't we have the same behavior as local modules where "absolute path is resolved relative to context" and "relative path is resolved relative to module source root" here for git dirs?

Otherwise you'd get different behavior for a given context dir path depending on whether you invoked the module from git or if you checked out the git repo locally and invoked the module locally, right?

Reply:
I'm not sure about that, if you use a gitsource, the  URL is the context dir of that source and it would also be the URL of the module.

`AsGitSource` doesn't provide any method to get a specific source root or anything, so I'm not sure how I could handle that